text/0000-edition-2021.md

```diff
@@ -0,0 +1,289 @@
+
+## Default edition for new projects
+
+The default edition for new projects created within cargo or other tooling will be 2021.
+
+## RFCs that propose edition changes
+
+RFCs that propose migrations should include details about how the migration between editions will work. This migration should be described in sufficient detail that the relevant teams can assess the feasibility of the migration. It will often make sense to consult the compiler team on this question specifically.
+
+* To perform this evaluation, an RFC proposing a migration should enumerate the situations that will no longer compile with the change. 
+* For each such situation, it should describe what suggestions will be made to modify the user's code such that it works on both the old and new edition. Optionally, the description may include "idiom lints" that run only in the new edition in order to make the code more idiomatic.
+    * Alternatively, if that scenario is deemed unlikely, the RFC can state that this sort of code will not be automatically ported between editions. It should then describe if it is possible to at least issue a warning that the code will no longer work or will change meaning in the new edition.
+
+## Tooling workflow
+
+The expected workflow for upgrading to a new edition is as follows:
+
+* **Prepare to upgrade:** Run `cargo fix --edition`: This will automatically prepare your code to be upgraded to next edition by applying any suggestions from the migration lints.
+    * For example, if your code is in Rust 2018, this will prepare you to move to the 2021 edition.
+    * Note that `cargo fix` does not actually move your code to the new edition. Instead, it produces code that works in both the old and the new edition.
+        * This allows people to upgrade and fix things in a "piecemeal" fashion. Because of this, however, the code produced by these suggestions is not always the most idiomatic, as it is not able to take advantage of features from the new edition.
+    * The expectation is that `cargo fix` should be able to fix the majority of crates out there, but it is not required that the tooling is able to handle every case. As long as code does not occur frequently in the wild, it is acceptable for the automated suggestions to be inapplicable to edge cases. The metrics section in this RFC includes some guidelines for how to measure this.
+* **Upgrade:** Edit your cargo.toml to include `edition = "2021"` instead of the older edition.
+    * The code should still build, but it may be necessary to make some fixes or other changes.
+* **Cleanup:** After upgrading, you should run `cargo fix` again. This second step will apply any remaining changes that are only possible in the new edition.
```

Comment:
Not sure if it is worth pointing out, but I think the intent here is to use the `--edition-idioms` flag at this step.

Reply:
This is something that @Manishearth and I have been discussing a bit on Zulip. I am not entirely convinced whether 'edition idioms' as a distinct category makes sense, or whether those lints are simply "style lints" that trigger on Rust 2021 code.

Reply:
It'd be good to discuss this a bit more on Zulip, I'd like to go over some of the examples of idiom lints I guess and think about it.